File: src/main/java/com/qlm/zombie/feature/FastCraftingFeature.java
/*
 * Copyright (c) 2026 QLM Zombie Mod
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * ----------------------------------------------------------------------------
 * Open Source Attribution:
 *   Original design inspired by FastWorkbench (https://github.com/Shadows-of-Fire/FastWorkbench)
 *   Copyright (c) Shadows-of-Fire. Licensed under MIT.
 *   Original design inspired by FastSuite (https://github.com/Shadows-of-Fire/FastSuite)
 *   This is an ORIGINAL Forge 1.20.1 implementation, NO code copied.
 * ----------------------------------------------------------------------------
 * 功能：工作台/锻造台/切石机合成使用缓存，避免每次重新扫描配方
 */
package com.qlm.zombie.feature;

import com.qlm.zombie.QLMZombieMod;
import net.minecraft.resources.ResourceLocation;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraft.world.Container;
import net.minecraft.world.inventory.CraftingContainer;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.CraftingRecipe;
import net.minecraft.world.item.crafting.Recipe;
import net.minecraft.world.item.crafting.RecipeType;
import net.minecraft.world.level.Level;
import net.minecraftforge.event.AnvilUpdateEvent;
import net.minecraftforge.event.entity.player.PlayerEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;

import java.util.HashMap;
import java.util.IdentityHashMap;
import java.util.Optional;

@Mod.EventBusSubscriber(modid = QLMZombieMod.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
public class FastCraftingFeature {

    private FastCraftingFeature() {}

    // 工作台配方缓存: Level -> (输入签名 -> 配方 + 结果)
    private static final IdentityHashMap<Level, HashMap<Long, CachedCraftingResult>> CRAFT_CACHE = new IdentityHashMap<>();
    private static final int MAX_CACHE = 512;

    private static final class CachedCraftingResult {
        CraftingRecipe recipe;
        ItemStack result;
    }

    /**
     * 工作台配方缓存查找（公共 API，可从 Mixin 或事件调用）。
     * 输入签名 = 容器内所有 ItemStack 的 id hash 聚合。
     */
    public static Optional<CraftingRecipe> findCraftingRecipe(CraftingContainer input, Level level) {
        if (level.isClientSide) return level.getRecipeManager().getRecipeFor(RecipeType.CRAFTING, input, level);

        long sig = signature(input);
        HashMap<Long, CachedCraftingResult> levelCache = CRAFT_CACHE.computeIfAbsent(level, lv -> new HashMap<>());

        CachedCraftingResult cached = levelCache.get(sig);
        // 缓存命中：仍需要 matches（因为 NBT 相同 Item 不代表输入布局相同；简化版：直接信任签名 + 重新验证一下输出）
        if (cached != null && cached.recipe.matches(input, level)) {
            return Optional.of(cached.recipe);
        }

        // 未命中，走原版查询
        Optional<CraftingRecipe> found = level.getRecipeManager().getRecipeFor(RecipeType.CRAFTING, input, level);
        found.ifPresent(recipe -> {
            if (levelCache.size() > MAX_CACHE) levelCache.clear();
            CachedCraftingResult cr = new CachedCraftingResult();
            cr.recipe = recipe;
            cr.result = recipe.assemble(input, level.registryAccess());
            levelCache.put(sig, cr);
        });

        return found;
    }

    /** 获取工作台缓存的合成结果 */
    public static ItemStack getCachedResult(CraftingRecipe recipe, CraftingContainer input, Level level) {
        return recipe.assemble(input, level.registryAccess());
    }

    // 根据容器内的 Item 生成 long 签名
    private static long signature(Container c) {
        long h = 1125899906842597L;
        for (int i = 0; i < c.getContainerSize(); i++) {
            ItemStack s = c.getItem(i);
            if (!s.isEmpty()) {
                ResourceLocation id = ForgeRegistries.ITEMS.getKey(s.getItem());
                int idHash = id != null ? id.hashCode() : 0;
                h = h * 31 + idHash;
                h = h * 31 + s.getCount();
            } else {
                h = h * 31 + 0xDEAD;
            }
        }
        return h;
    }

    @SubscribeEvent
    public static void onCrafted(PlayerEvent.ItemCraftedEvent event) {
        // 预留：可以在这里统计性能、或强制清理 cache。
        // 当前版本不需要特殊处理，因为签名哈希已经把输入布局编码到 key 中。
    }
}

File: src/main/kotlin/com/qlm/zombie/feature/FastCraftingFeature.kt
package com.qlm.zombie.feature

import com.qlm.zombie.QLMZombieMod
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.CraftingContainer
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.CraftingRecipe
import net.minecraft.world.item.crafting.RecipeType
import net.minecraftforge.event.entity.player.PlayerEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod

@Mod.EventBusSubscriber(modid = QLMZombieMod.MOD_ID, bus = Mod.EventBusSubscriber.Bus.FORGE)
object FastCraftingFeature {

    @SubscribeEvent
    fun onItemCrafted(event: PlayerEvent.ItemCraftedEvent) {
        val player = event.entity
        val resultStack = event.crafting
        val crafting = event.inventory as? CraftingContainer ?: return

        if (resultStack.isEmpty) return
        if (player.level().isClientSide) return

        val recipe = player.level().recipeManager.getRecipeFor(
            RecipeType.CRAFTING,
            crafting,
            player.level()
        ).orElse(null) as? CraftingRecipe ?: return

        if (!recipe.canCraftInDimensions(3, 3) && !recipe.canCraftInDimensions(2, 2)) return

        tryQuickCraft(player, crafting, resultStack, recipe)
    }

    private fun tryQuickCraft(
        player: Player,
        crafting: CraftingContainer,
        originalResult: ItemStack,
        recipe: CraftingRecipe
    ) {
        val inventory = player.inventory
        var craftedCount = 0
        val registryAccess = player.level().registryAccess()

        while (recipe.matches(crafting, player.level()) && craftedCount < 64) {
            val remainingStacks = recipe.getRemainingItems(crafting)

            for (i in 0 until crafting.containerSize) {
                val currentStack = crafting.getItem(i)
                val remainingStack = if (i < remainingStacks.size) remainingStacks[i] else ItemStack.EMPTY

                if (!currentStack.isEmpty) {
                    currentStack.shrink(1)
                    if (currentStack.isEmpty && !remainingStack.isEmpty) {
                        crafting.setItem(i, remainingStack)
                    }
                } else if (!remainingStack.isEmpty) {
                    crafting.setItem(i, remainingStack)
                }
            }

            val newResult = recipe.assemble(crafting, registryAccess)
            if (newResult.isEmpty) break

            if (!inventory.add(newResult.copy())) {
                player.drop(newResult, false)
                break
            }

            craftedCount++
        }

        if (craftedCount > 0 && player is ServerPlayer) {
            QLMZombieMod.LOGGER.debug(
                "[FastCraft] 玩家 {} 快速制作 {} x{}",
                player.name.string,
                originalResult.item.descriptionId,
                craftedCount
            )
        }
    }
}
